tokio/docs/runtime-stats.md

```diff
@@ -0,0 +1,137 @@
+
+```rust=
+// Handle that can be acquired via `Runtime` or via `Handle::current`.
+let handle = Handle::current();
+let stats = handle.stats();
+
+// executor level stats.
+let executor = stats.executor();
+
+// per-worker stats via the executor.
+for worker in executor.workers() {
+    println!("{}", worker.steal_count());
+}
+
+println!("{}", executor.global_queue_depth());
+
+// io driver stats.
+let io = stats.io();
+
+println!("{}", io.sources_count());
+
+let timer = stats.timer();
+
+println!("{}", timer.entry_count());
+```
+
+### Executor
+
+The handle will provide stats for each worker thread or in the case of the single threaded executor will provide a single worker. This provides a detailed view into what each worker is doing. Providing the ability for the `tokio-metrics` crate to expose the stats aggregated as a single metric or as a per-worker metric.
+
+The values will be updated in batch from the executor to avoid needing to stream the data on every action. This should amortize the cost by only needing to emit stats at a specific executor wall clock time. Where the executor wall clock time is determinetd by a single executor tick rather than actual system time. This allows the collectors to observe the time and the stats to determine how long certain executor cycles took. This removes the need to acquire the time during executor cycles.
+
+Each worker will expose these stats, updated in batches:
+
+- Amount of futures executed
+- Amount of work steal attempts
+- Amount of parks
+- Amount of times local queue overflowed into the global queue
+- Max local queue depth
+- Min local queue depth
+- Avg local queue depth
+- Queue depth at time of batch emission
+- Amount of executor ticks (loop iterations)
+- Number of `block_in_place` tasks entered
+
+The main goal of this implementation is to allow a user to run this metrics collection at all times in production with minimal overhead to their application. This would allow users to alarm on any regressions and track how the runtime is performing.
+
+Some of the stats include min/max (specifically the queue depth stats) this is because the depth of the queues changes throughout the stats batch window. The value could start low, spike up during the middle of the window then come back down. To understand this behavior the executor stats module will aggregate the depth values to reduce the need to stream the values.
+
+### Global queue
+
+Since, the global inject queue is not attached to a single worker its length value can be sampled directly from collector thread. In addition, the global inject queue length is already tracked via an `AtomicUsize`, therefore this value can just be exposed via the `Stats` struct directly.
```

Comment:
nit: this is the first mention of a `Stats` struct. It's also the first mention of the fact that the global inject queue length is tracked, and of the use of `AtomicUsize`.